### DISCUSSION TITLE
[A-L] (18 giugno 2024) Esercizio 2

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/progettazione-di-algoritmi#92] [Esami A-L]
Dare lo pseudocodice di un algoritmo che prenda in input il vettore dei padri di un albero $\Large P []$ e trovi l'elenco dei vertici che non sono foglie nell'albero.

L'algoritmo dovrebbe avere una complessità $\Large O(n)$.

**Comment by FeddyLix17** [comment]
```python

def es2(P):

    # tengo traccia del numero di figli di ogni vertice
    # utilizzando un array ausiliario
    num_figli = [0] * len(P)

    # per ogni vertice dell'albero
    for nodo in range(1, len(P)):
        
        # ricavo il padre tramite l'array passato in input
        padre = P[nodo]
        
        # se presente
        if padre is not None:
            
            # incremento il suo numero di figli 
            num_figli[padre] += 1
     
     # tutti i vertici con almeno un figlio
     # allo stesso tempo non saranno foglie
     not_foglie = []
     for nodo in range(1, len(num_figli)):    # len(num_figli) == len(padri))
         if num_figli[nodo] > 0:
             not_foglie.append(nodo)

    return not_foglie
```

I 2 cicli for hanno entrambi costo O(n), rispettando la complessità dell'algoritmo richiesta dall'esercizio.

**Comment by cornflexxx** [comment]
Dato che l'array $P$ dei padri ha come unici elementi vertici che hanno almeno un figlio, è sufficiente scorrere su $P$ e prendere tutti i vertici distinti.

$$Sol := \\{P[i] \text{ con } 0\le i\le len(P)\\}$$

**Comment by Elia-Belli** [comment]
Per definizione l'array dei padri $P$ contiene solo nodi che sono padri di qualcuno e quindi non foglie.
```python
def not_foglie(P):
    Sol = {}

    for i in range(len(P)):      #O(n)
	Sol.add(P[i])       

    return Sol
```

**Comment by CiottoloMaggico** [comment]
Un nodo può essere padre di più figli, occorrendo più volte all'interno del vettore dei padri. Per risolvere il problema in O(n) utilizzando un semplice array, prima segno i nodi non foglie e poi li trasferisco in un array risultato.
![immagine](https://github.com/user-attachments/assets/adc0e8bb-127e-4b18-9a9b-ad7b10f53c3e)
